### PR TITLE
Get OverloadType PETSc.vec

### PR DESCRIPTION
**Pull request by Ig-dolci**
This PR adds an `OverloadType` method to get the `PETSc.petsc4py.Vec` object.

### CODE REVIEW

File: pyadjoint/overloaded_type.py [hunk 1]
```diff
@@ -320,6 +320,28 @@ def _ad_str(self):
         """
         return str(self)
 
+    def _ad_petsc_vec_read_only(self):
+        """This method must be overridden.
```
Comment (dham) [inline pyadjoint/overloaded_type.py:324]: I think that in this case this overstates the case.
```suggestion
        """This method should be overwritten by types which can return a PETSc Vec.
```

File: pyadjoint/overloaded_type.py [hunk 2]
```diff
@@ -320,6 +320,28 @@ def _ad_str(self):
...
+        The method should implement a routine to return a read only a PETSc Vec object.
+
+        Returns:
+            PETSc.Vec: A PETSc Vec object containing the data of the overloaded.
+
+        """
+        raise NotImplementedError
+
+    def _ad_petsc_vec_write_only(self):
+        """This method must be overridden.
```
Comment (dham) [inline pyadjoint/overloaded_type.py:335]: as above.

File: pyadjoint/overloaded_type.py [hunk 3]
```diff
@@ -320,6 +320,28 @@ def _ad_str(self):
...
```
Comment (Ig-dolci) [inline pyadjoint/overloaded_type.py:335]:
```suggestion
        """This method should be overwritten by types which can return a PETSc Vec.
```

File: pyadjoint/overloaded_type.py (hunk 2 above)
Comment (Ig-dolci) [reply pyadjoint/overloaded_type.py]: done.

**Comment by jrmaddison** [comment]
Should document that these are context managers, and also document what is expected on entry/exit (ghost updates? other data consistency updates?).

**Comment by Ig-dolci** [comment]
> Should document that these are context managers, and also document what is expected on entry/exit (ghost updates? other data consistency updates?).

Yes. It is a good point. I will study more about the management of Petsc vecs and try to improve the docs.

File: pyadjoint/overloaded_type.py [hunk 4]
```diff
@@ -320,6 +320,42 @@ def _ad_str(self):
         """
         return str(self)
 
+    def _ad_petsc_vec_read_only(self):
```
Comment (Ig-dolci) [inline pyadjoint/overloaded_type.py:323]: Should I consider adding an optional argument like with_halos? Or does it not make sense?

**Comment by Ig-dolci** [comment]
This PR is not necessary anymore. OverloadedTypes to copy and update PETSc Vecs are available at the [PR 143](https://github.com/dolfin-adjoint/pyadjoint/pull/143).

**Pull request closed without merging**